File: mobile/apps/photos/lib/models/collection/collection_tree.dart
import "package:photos/models/collection/collection.dart";

/// Represents a node in the collection tree hierarchy
class CollectionTreeNode {
  final Collection collection;
  final List<CollectionTreeNode> children;
  final int depth;

  CollectionTreeNode({
    required this.collection,
    List<CollectionTreeNode>? children,
    this.depth = 0,
  }) : children = children ?? [];

  /// Returns true if this node is a root node (no parent)
  bool get isRoot => (collection.pubMagicMetadata.parentID ?? 0) == 0;

  /// Returns true if this node is a leaf node (no children)
  bool get isLeaf => children.isEmpty;

  /// Returns the parent ID, or 0 if this is a root node
  int get parentID => collection.pubMagicMetadata.parentID ?? 0;

  /// Returns the total count of descendants (children, grandchildren, etc.)
  int get descendantCount {
    int count = children.length;
    for (final child in children) {
      count += child.descendantCount;
    }
    return count;
  }

  /// Returns a flat list of all descendant collections
  List<Collection> get descendants {
    final List<Collection> result = [];
    for (final child in children) {
      result.add(child.collection);
      result.addAll(child.descendants);
    }
    return result;
  }

  /// Returns a flat list of all descendant collection IDs
  Set<int> get descendantIDs {
    final Set<int> result = {};
    for (final child in children) {
      result.add(child.collection.id);
      result.addAll(child.descendantIDs);
    }
    return result;
  }

  /// Checks if the given collection ID is a descendant of this node
  bool isDescendant(int collectionID) {
    return descendantIDs.contains(collectionID);
  }

  /// Adds a child node to this node
  void addChild(CollectionTreeNode child) {
    children.add(child);
  }

  /// Removes a child node from this node
  bool removeChild(int collectionID) {
    final initialLength = children.length;
    children.removeWhere((c) => c.collection.id == collectionID);
    return children.length < initialLength;
  }

  /// Finds a node by collection ID in this subtree
  CollectionTreeNode? findNode(int collectionID) {
    if (collection.id == collectionID) {
      return this;
    }
    for (final child in children) {
      final found = child.findNode(collectionID);
      if (found != null) {
        return found;
      }
    }
    return null;
  }
}

/// Represents the entire collection tree structure
class CollectionTree {
  final List<CollectionTreeNode> roots;
  final Map<int, CollectionTreeNode> _nodeMap;

  CollectionTree({required this.roots, Map<int, CollectionTreeNode>? nodeMap})
      : _nodeMap = nodeMap ?? {};

  /// Gets a node by collection ID
  CollectionTreeNode? getNode(int collectionID) {
    return _nodeMap[collectionID];
  }

  /// Gets all nodes as a flat list
  List<CollectionTreeNode> get allNodes => _nodeMap.values.toList();

  /// Gets all collections as a flat list
  List<Collection> get allCollections =>
      _nodeMap.values.map((node) => node.collection).toList();

  /// Gets the path from root to the given collection ID
  /// Returns null if collection is not found
  List<Collection>? getPath(int collectionID) {
    final node = _nodeMap[collectionID];
    if (node == null) return null;

    final List<Collection> path = [];
    CollectionTreeNode? current = node;

    while (current != null) {
      path.insert(0, current.collection);
      final parentID = current.parentID;
      if (parentID == 0) break;
      current = _nodeMap[parentID];
    }

    return path;
  }

  /// Gets breadcrumb trail for a collection (list of collection names from root)
  List<String> getBreadcrumbs(int collectionID) {
    final path = getPath(collectionID);
    if (path == null) return [];
    return path.map((c) => c.displayName).toList();
  }

  /// Checks if moving sourceID under targetID would create a cycle
  bool wouldCreateCycle(int sourceID, int targetID) {
    if (sourceID == targetID) return true;

    final sourceNode = _nodeMap[sourceID];
    if (sourceNode == null) return false;

    // Check if target is a descendant of source
    return sourceNode.isDescendant(targetID);
  }

  /// Gets the depth of a collection in the tree
  int getDepth(int collectionID) {
    final node = _nodeMap[collectionID];
    return node?.depth ?? 0;
  }
}

File: web/packages/new/photos/utils/collection-tree.ts
/**
 * Utilities for building and working with hierarchical collection trees.
 *
 * This module provides functionality to build a tree structure from flat
 * collections using their parent-child relationships stored in pubMagicMetadata.
 */

import type { Collection } from "ente-media/collection";

/**
 * Represents a node in the collection tree hierarchy.
 */
export interface CollectionTreeNode {
    /** The collection at this node */
    collection: Collection;
    /** Child nodes in the tree */
    children: CollectionTreeNode[];
    /** Depth of this node in the tree (0 for root) */
    depth: number;
}

/**
 * Represents the complete collection tree structure.
 */
export interface CollectionTree {
    /** Root nodes (collections with no parent) */
    roots: CollectionTreeNode[];
    /** Map from collection ID to node for fast lookups */
    nodeMap: Map<number, CollectionTreeNode>;
}

/**
 * Builds a collection tree from a flat list of collections.
 *
 * @param collections - Flat array of all collections
 * @returns The built collection tree
 */
export function buildCollectionTree(
    collections: Collection[],
): CollectionTree {
    const nodeMap = new Map<number, CollectionTreeNode>();
    const roots: CollectionTreeNode[];

    // Create nodes for all collections
    for (const collection of collections) {
        nodeMap.set(collection.id, {
            collection,
            children: [],
            depth: 0,
        });
    }

    // Build parent-child relationships
    const rootNodes: CollectionTreeNode[] = [];
    for (const collection of collections) {
        const node = nodeMap.get(collection.id)!;
        const parentID = collection.pubMagicMetadata?.data.parentID;

        if (!parentID || parentID === 0) {
            rootNodes.push(node);
        } else {
            const parent = nodeMap.get(parentID);
            if (parent) {
                parent.children.push(node);
            } else {
                // Parent not found, treat as root
                rootNodes.push(node);
            }
        }
    }

    // Calculate depths
    function calculateDepth(node: CollectionTreeNode, depth: number): void {
        node.depth = depth;
        for (const child of node.children) {
            calculateDepth(child, depth + 1);
        }
    }

    for (const root of rootNodes) {
        calculateDepth(root, 0);
    }

    return {
        roots: rootNodes,
        nodeMap,
    };
}

/**
 * Gets the path from root to a specific collection.
 *
 * @param tree - The collection tree
 * @param collectionID - The target collection ID
 * @returns Array of collections from root to target, or undefined if not found
 */
export function getCollectionPath(
    tree: CollectionTree,
    collectionID: number,
): Collection[] | undefined {
    const node = tree.nodeMap.get(collectionID);
    if (!node) return undefined;

    const path: Collection[] = [];
    let current: CollectionTreeNode | undefined = node;

    while (current) {
        path.unshift(current.collection);
        const parentID = current.collection.pubMagicMetadata?.data.parentID;
        if (!parentID || parentID === 0) break;
        current = tree.nodeMap.get(parentID);
    }

    return path;
}

/**
 * Gets breadcrumb names from root to a specific collection.
 *
 * @param tree - The collection tree
 * @param collectionID - The target collection ID
 * @returns Array of collection names from root to target
 */
export function getCollectionBreadcrumbs(
    tree: CollectionTree,
    collectionID: number,
): string[] {
    const path = getCollectionPath(tree, collectionID);
    return path ? path.map((c) => c.name) : [];
}

/**
 * Gets all descendant collections of a given collection.
 *
 * @param tree - The collection tree
 * @param collectionID - The parent collection ID
 * @returns Array of all descendant collections
 */
export function getDescendants(
    tree: CollectionTree,
    collectionID: number,
): Collection[] {
    const node = tree.nodeMap.get(collectionID);
    if (!node) return [];

    const descendants: Collection[] = [];

    function collectDescendants(n: CollectionTreeNode): void {
        for (const child of n.children) {
            descendants.push(child.collection);
            collectDescendants(child);
        }
    }

    collectDescendants(node);
    return descendants;
}

/**
 * Gets immediate children of a collection.
 *
 * @param tree - The collection tree
 * @param collectionID - The parent collection ID
 * @returns Array of immediate child collections
 */
export function getChildren(
    tree: CollectionTree,
    collectionID: number,
): Collection[] {
    const node = tree.nodeMap.get(collectionID);
    return node ? node.children.map((n) => n.collection) : [];
}

/**
 * Checks if moving a collection would create a cycle.
 *
 * @param tree - The collection tree
 * @param sourceID - The collection being moved
 * @param targetID - The new parent collection
 * @returns true if this would create a cycle
 */
export function wouldCreateCycle(
    tree: CollectionTree,
    sourceID: number,
    targetID: number,
): boolean {
    if (sourceID === targetID) return true;

    const sourceNode = tree.nodeMap.get(sourceID);
    if (!sourceNode) return false;

    // Check if target is a descendant of source
    const descendants = getDescendants(tree, sourceID);
    return descendants.some((d) => d.id === targetID);
}

/**
 * Gets the depth of a collection in the tree.
 *
 * @param tree - The collection tree
 * @param collectionID - The collection ID
 * @returns The depth (0 for root, -1 if not found)
 */
export function getCollectionDepth(
    tree: CollectionTree,
    collectionID: number,
): number {
    const node = tree.nodeMap.get(collectionID);
    return node ? node.depth : -1;
}

/**
 * Finds the maximum depth of descendants for a given node.
 *
 * @param node - The starting node
 * @returns The maximum depth of any descendant relative to this node
 */
export function getMaxDescendantDepth(node: CollectionTreeNode): number {
    if (node.children.length === 0) return 0;

    let maxDepth = 0;
    for (const child of node.children) {
        const childDepth = 1 + getMaxDescendantDepth(child);
        maxDepth = Math.max(maxDepth, childDepth);
    }
    return maxDepth;
}

/**
 * Sorts collections to maintain tree order (parents before children).
 *
 * @param collections - Collections to sort
 * @returns Sorted collections
 */
export function sortCollectionsTreeOrder(
    collections: Collection[],
): Collection[] {
    const tree = buildCollectionTree(collections);
    const sorted: Collection[] = [];

    function traverse(node: CollectionTreeNode): void {
        sorted.push(node.collection);
        for (const child of node.children) {
            traverse(child);
        }
    }

    for (const root of tree.roots) {
        traverse(root);
    }

    return sorted;
}
